common/ferrostar/src/models.rs

```diff
@@ -97,13 +97,34 @@ impl From<GeographicCoordinate> for Point {
 /// and are used for recalculating when the user deviates from the expected route.
 ///
 /// Note that support for properties beyond basic geographic coordinates varies by routing engine.
-#[derive(Clone, Copy, PartialEq, PartialOrd, Debug, Serialize, Deserialize)]
+#[derive(Clone, PartialEq, PartialOrd, Debug, Serialize, Deserialize)]
 #[cfg_attr(feature = "uniffi", derive(uniffi::Record))]
 #[cfg_attr(feature = "wasm-bindgen", derive(Tsify))]
 #[cfg_attr(feature = "wasm-bindgen", tsify(into_wasm_abi, from_wasm_abi))]
 pub struct Waypoint {
     pub coordinate: GeographicCoordinate,
     pub kind: WaypointKind,
+    /// Optional additional properties that will be passed on to the [`crate::routing_adapters::RouteRequestGenerator`].
+    ///
+    /// Most users should prefer convenience functions like [`Waypoint::new_with_valhalla_properties`]
+    /// (or, on platforms like iOS and Android with UniFFI bindings, [`crate::routing_adapters::valhalla::create_waypoint_with_valhalla_properties`]).
+    ///
+    /// # Format guidelines
+    ///
+    /// This MAY be any format agreed upon by both the request generator and response parser.
+    /// However, to promote interoperability, all implementations in the Ferrostar codebase
+    /// MUST use JSON.
+    ///
+    /// We selected JSON is selected not because it is good,
+    /// but because generics (i.e., becoming `Waypoint<T>`, where an example `T` is `ValhallaProperties`)
+    /// would be way too painful, particularly for foreign code.
+    /// Especially JavaScript.
+    ///
+    /// In any case, [`crate::routing_adapters::RouteRequestGenerator`] and [`crate::routing_adapters::RouteResponseParser`]
+    /// implementations SHOULD document their level support for this,
+    /// ideally with an exportable record type.
+    #[cfg_attr(feature = "uniffi", uniffi(default))]
```

Comment:
I discovered a new UniFFI feature 🎉

This lets us mark a property of a record type (struct) as optional. At the FFI level, we get much easier constructors. This is especially noticeable in Swift.